File: java/src/tarehart/rlbot/physics/DistancePlot.java
package tarehart.rlbot.physics;

import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.DistanceTimeSpeed;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.carpredict.AccelerationModel;
import tarehart.rlbot.intercept.StrikeProfile;
import tarehart.rlbot.time.Duration;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class DistancePlot {

    ArrayList<DistanceTimeSpeed> plot = new ArrayList<>();

    public DistancePlot(DistanceTimeSpeed start) {
        plot.add(start);
    }

    public void addSlice(DistanceTimeSpeed dts) {
        plot.add(dts);
    }

    public List<DistanceTimeSpeed> getSlices() {
        return plot;
    }

    public Optional<DistanceTimeSpeed> getMotionAfterDuration(Duration time) {
        if (time.compareTo(plot.get(0).getTime()) < 0 || time.compareTo(plot.get(plot.size() - 1).getTime()) > 0) {
            return Optional.empty();
        }

        for (int i = 0; i < plot.size() - 1; i++) {
            DistanceTimeSpeed current = plot.get(i);
            DistanceTimeSpeed next = plot.get(i + 1);
            if (next.getTime().compareTo(time) > 0) {

                double simulationStepSeconds = next.getTime().getSeconds() - current.getTime().getSeconds();
                double tweenPoint = (time.getSeconds() - current.getTime().getSeconds()) / simulationStepSeconds;
                double distance = (1 - tweenPoint) * current.getDistance() + tweenPoint * next.getDistance();
                double speed = (1 - tweenPoint) * current.getSpeed() + tweenPoint * next.getSpeed();
                return Optional.of(new DistanceTimeSpeed(distance, time, speed));
            }
        }

        return Optional.of(plot.get(plot.size() - 1));
    }

    public Optional<DistanceTimeSpeed> getMotionAfterDistance(double distance) {

        for (int i = 0; i < plot.size() - 1; i++) {
            DistanceTimeSpeed current = plot.get(i);
            DistanceTimeSpeed next = plot.get(i + 1);
            if (next.getDistance() > distance) {
                double simulationStepSeconds = next.getTime().getSeconds() - current.getTime().getSeconds();
                double tweenPoint = (distance - current.getDistance()) / (next.getDistance() - current.getDistance());
                Duration moment = current.getTime().plusSeconds(simulationStepSeconds * tweenPoint);
                double speed = (1 - tweenPoint) * current.getSpeed() + tweenPoint * next.getSpeed();
                return Optional.of(new DistanceTimeSpeed(distance, moment, speed));
            }
        }
        return Optional.empty();
    }

    public Optional<Duration> getTravelTime(double distance) {
        Optional<DistanceTimeSpeed> motionAt = getMotionAfterDistance(distance);
        return motionAt.map(DistanceTimeSpeed::getTime);
    }

    public Optional<DistanceTimeSpeed> getMotionUponArrival(CarData carData, Vector3 destination, StrikeProfile strikeProfile) {

        double orientSeconds = AccelerationModel.getSteerPenaltySeconds(carData, destination) + strikeProfile.maneuverSeconds;
        double distance = carData.position.flatten().distance(destination.flatten());

        return getMotionAfterDistance(distance).map(dts -> new DistanceTimeSpeed(dts.getDistance(), dts.getTime().plusSeconds(orientSeconds), dts.getSpeed()));

        // TODO: incorporate the speedup from the strike profile.
    }

    public Optional<DistanceTimeSpeed> getMotionAfterDuration(CarData carData, Vector3 target, Duration time, StrikeProfile strikeProfile) {

        double orientSeconds = AccelerationModel.getSteerPenaltySeconds(carData, target) + strikeProfile.maneuverSeconds;

        double totalSeconds = time.getSeconds();
        double secondsSpentAccelerating = Math.max(0, totalSeconds - orientSeconds);

        if (strikeProfile.dodgeSeconds == 0 || strikeProfile.speedBoost == 0) {
            Optional<DistanceTimeSpeed> motion = getMotionAfterDuration(Duration.ofSeconds(secondsSpentAccelerating));
            return motion.map(dts -> new DistanceTimeSpeed(dts.getDistance(), time, dts.getSpeed()));
        }

        double speedupSeconds = strikeProfile.dodgeSeconds;
        double speedBoost = strikeProfile.speedBoost;
        if (secondsSpentAccelerating < speedupSeconds) {
            // Not enough time for a full strike.
            double beginningSpeed = plot.get(0).getSpeed();
            double increasedSpeed = Math.min(beginningSpeed + speedBoost, AccelerationModel.SUPERSONIC_SPEED);
            return Optional.of(new DistanceTimeSpeed(increasedSpeed * secondsSpentAccelerating, time, increasedSpeed));
        }

        double accelSecondsBeforeStrike = secondsSpentAccelerating - speedupSeconds;
        Optional<DistanceTimeSpeed> dtsOption = getMotionAfterDuration(Duration.ofSeconds(accelSecondsBeforeStrike));

        if (dtsOption.isPresent()) {
            DistanceTimeSpeed dts = dtsOption.get();
            double beginningSpeed = dts.getSpeed();
            double increasedSpeed = Math.min(beginningSpeed + speedBoost, AccelerationModel.SUPERSONIC_SPEED);
            return Optional.of(new DistanceTimeSpeed(dts.getDistance() + increasedSpeed * speedupSeconds, time, increasedSpeed));
        } else {
            // We ran out of data in the distance plot.
            return Optional.empty();
        }
    }
}

File: java/src/tarehart/rlbot/physics/DistancePlot.kt
package tarehart.rlbot.physics

import tarehart.rlbot.input.CarData
import tarehart.rlbot.math.DistanceTimeSpeed
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.carpredict.AccelerationModel
import tarehart.rlbot.intercept.StrikeProfile
import tarehart.rlbot.time.Duration

import java.util.ArrayList
import java.util.Optional

class DistancePlot(start: DistanceTimeSpeed) {

    private var plot = ArrayList<DistanceTimeSpeed>()

    val slices: List<DistanceTimeSpeed>
        get() = plot

    init {
        plot.add(start)
    }

    fun addSlice(dts: DistanceTimeSpeed) {
        plot.add(dts)
    }

    fun getEndPoint(): DistanceTimeSpeed {
        return plot[plot.size - 1]
    }

    fun getMotionAfterDuration(time: Duration): Optional<DistanceTimeSpeed> {
        if (time < plot[0].time || time > plot[plot.size - 1].time) {
            return Optional.empty()
        }

        for (i in 0 until plot.size - 1) {
            val current = plot[i]
            val next = plot[i + 1]
            if (next.time > time) {

                val simulationStepSeconds = next.time.seconds - current.time.seconds
                val tweenPoint = (time.seconds - current.time.seconds) / simulationStepSeconds
                val distance = (1 - tweenPoint) * current.distance + tweenPoint * next.distance
                val speed = (1 - tweenPoint) * current.speed + tweenPoint * next.speed
                return Optional.of(DistanceTimeSpeed(distance, time, speed))
            }
        }

        return Optional.of(plot[plot.size - 1])
    }

    fun getMotionAfterDistance(distance: Double): Optional<DistanceTimeSpeed> {

        for (i in 0 until plot.size - 1) {
            val current = plot[i]
            val next = plot[i + 1]
            if (next.distance > distance) {
                val simulationStepSeconds = next.time.seconds - current.time.seconds
                val tweenPoint = (distance - current.distance) / (next.distance - current.distance)
                val moment = current.time.plusSeconds(simulationStepSeconds * tweenPoint)
                val speed = (1 - tweenPoint) * current.speed + tweenPoint * next.speed
                return Optional.of(DistanceTimeSpeed(distance, moment, speed))
            }
        }
        return Optional.empty()
    }

    fun getTravelTime(distance: Double): Optional<Duration> {
        val motionAt = getMotionAfterDistance(distance)
        return motionAt.map { it.time }
    }

    fun getMotionUponArrival(carData: CarData, destination: Vector3, strikeProfile: StrikeProfile): Optional<DistanceTimeSpeed> {

        val orientSeconds = AccelerationModel.getSteerPenaltySeconds(carData, destination) + strikeProfile.maneuverSeconds
        val distance = carData.position.flatten().distance(destination.flatten())

        return getMotionAfterDistance(distance).map { DistanceTimeSpeed(it.distance, it.time.plusSeconds(orientSeconds), it.speed) }

        // TODO: incorporate the speedup from the strike profile.
    }

    fun getMotionAfterDuration(carData: CarData, target: Vector3, time: Duration, strikeProfile: StrikeProfile): Optional<DistanceTimeSpeed> {

        val orientSeconds = AccelerationModel.getSteerPenaltySeconds(carData, target) + strikeProfile.maneuverSeconds

        val totalSeconds = time.seconds
        val secondsSpentAccelerating = Math.max(0.0, totalSeconds - orientSeconds)

        if (strikeProfile.dodgeSeconds == 0.0 || strikeProfile.speedBoost == 0.0) {
            val motion = getMotionAfterDuration(Duration.ofSeconds(secondsSpentAccelerating))
            return motion.map { DistanceTimeSpeed(it.distance, time, it.speed) }
        }

        val speedupSeconds = strikeProfile.dodgeSeconds
        val speedBoost = strikeProfile.speedBoost
        if (secondsSpentAccelerating < speedupSeconds) {
            // Not enough time for a full strike.
            val beginningSpeed = plot[0].speed
            val increasedSpeed = Math.min(beginningSpeed + speedBoost, AccelerationModel.SUPERSONIC_SPEED)
            return Optional.of(DistanceTimeSpeed(increasedSpeed * secondsSpentAccelerating, time, increasedSpeed))
        }

        val accelSecondsBeforeStrike = secondsSpentAccelerating - speedupSeconds
        val dtsOption = getMotionAfterDuration(Duration.ofSeconds(accelSecondsBeforeStrike))

        return dtsOption.map {
            val beginningSpeed = it.speed
            val increasedSpeed = Math.min(beginningSpeed + speedBoost, AccelerationModel.SUPERSONIC_SPEED)
            DistanceTimeSpeed(it.distance + increasedSpeed * speedupSeconds, time, increasedSpeed)
        }
    }
}
